aptos-move/aptos-aggregator/src/aggregator_extension.rs

```diff
@@ -492,8 +485,115 @@ impl AggregatorData {
         Ok(snapshot_id)
     }
 
-    pub fn read_snapshot(&self, _id: AggregatorVersionedID) -> PartialVMResult<u128> {
-        unimplemented!();
+    pub fn create_new_snapshot(&mut self, id: AggregatorID, value: SnapshotValue) {
+        let snapshot_state = AggregatorSnapshotState::Create { value };
+
+        self.aggregator_snapshots.insert(id, AggregatorSnapshot {
+            id,
+            state: snapshot_state,
+        });
+    }
+
+    /// Returns a mutable reference to an aggregator snapshot with `id`.
+    /// If transaction that is currently executing did not initialize it, a new aggregator snapshot instance is created.
+    /// Note: when we say "aggregator snapshot instance" here we refer to Rust struct and
+    /// not to the Move aggregator snapshot.
+    pub fn get_snapshot(
+        &mut self,
+        id: AggregatorID,
+        resolver: &dyn AggregatorResolver,
+    ) -> PartialVMResult<&AggregatorSnapshot> {
+        let snapshot = match self.aggregator_snapshots.entry(id) {
+            Entry::Vacant(entry) => {
+                // Otherwise, we have to go to storage and read the value.
+                let value_from_storage = resolver
+                    .get_aggregator_v2_value(&id, AggregatorReadMode::Aggregated)
+                    .map_err(|e| {
+                        extension_error(format!(
+                            "Could not find the value of the aggregator: {}",
+                            e
+                        ))
+                    })?;
+                entry.insert(AggregatorSnapshot {
+                    id,
+                    state: AggregatorSnapshotState::Reference {
+                        speculative_value: SnapshotValue::try_from(value_from_storage)?,
+                    },
+                })
+            },
+            Entry::Occupied(entry) => entry.into_mut(),
+        };
+        Ok(snapshot)
+    }
+
```

Comment:
in general, I feel that snapshot APIs should go into their own file, it is very hard to navigate in the code right now